### PR TITLE
Encode streamed audio with one encoder per stream

### LINKED ISSUE #11733: The audio stream received by the Audio component is not continuous
### Describe the bug

I use `gr.Audio(label="Stream Output Audio", streaming=True,format='wav')` to receive audio stream,but the audio sounds discontinuous and choppy.

Image

You can easily reproduce this bug using my code. Input a audio and listen the received audio.

### Have you searched existing issues?  🔎

- [x] I have searched and found no existing issues

### Reproduction

```python
import gradio as gr

def synthesize(audio):
    yield None,None
    sr,data=audio
    step=4000
    index=0
    if len(data.shape)==2:
        data=data[:,0]
    assert(1==len(data.shape))
    length=data.shape[0]
    while index<length:
        cur_sent=data[index:index+step]
        index+=step
        if index<length:
            yield (sr,cur_sent),None
        else:
            yield (sr,cur_sent),(sr,data)
        
    

def main():
    description = ("test audio receive stream")
    inputs = [
        gr.Audio(label="audio input",sources=['upload'])
    ]


    outputs = [gr.Audio(label="Stream Output Audio", streaming=True,format='wav'),gr.Audio(label="Output Audio", streaming=False,format='wav')]

    gr.Interface(fn=synthesize,
                 description=description,
                 inputs=inputs,
                 outputs=outputs,
                 title="example",
                 cache_examples=False,
                 ).launch(share=False, server_port=8000)


if __name__ == "__main__":
    main()
```

### Screenshot

_No response_

### Logs

```shell

```

### System Info

```shell
gradio  5.42.0
Python 3.12.11
```

### Severity

I can work around it

### PR DESCRIPTION
**Pull request by hysts**
## Description

`gr.Audio(streaming=True)` encoded every yielded chunk into its own AAC file. Each of those files opens with an encoder priming frame that has no overlap-add partner, so the player rendered it as a short silence and the stream broke up at every chunk boundary.

Decoding what the server sends for an 8.000 s, 16 kHz signal fed in 4000-sample chunks, before and after:

| | decoded | gaps | avg gap | silence | size |
|---|---|---|---|---|---|
| `main` | 10.240 s | 33 | 67.9 ms | 21.9% | 70944 B |
| this PR | 8.064 s | 1, at t=0 | 64.1 ms | 0.8% | 31760 B |

This keeps one ffmpeg encoder alive for the length of a stream and cuts the HLS segments out of its continuous output, so consecutive segments share encoder state and join up. The one remaining gap is the stream's own priming frame, which has nothing before it to join to. Replaying the segments in Chromium, the per-chunk gaps are gone.

The same thing shows up as a length: 33 gaps of 67.9 ms is 2.24 s, which is exactly how much longer than its 8.000 s source the stream decodes. The playlist does not account for it either, because `#EXTINF` carries each source chunk's duration, so it sums to 8.000 s while the media it describes is 10.240 s. After this change the two agree, 8.064 s against 8.064 s, which is what the new backend test asserts.

Things worth knowing about the implementation:

- The encoder command carries `-probesize 32 -analyzeduration 0`. Without those, ffmpeg buffers before it emits anything and the first segment arrives about two seconds late, which would trade the gaps for a startup delay. Output is byte-identical either way.
- The encoder cannot live on the component, which is shared across sessions, or on the `MediaStream`, which is created after `stream_output` returns, so it lives in a module-level dict keyed by the stream's playlist path.
- Teardown runs from `end_stream()`, and the paths that reach it are normal completion, the exception handler in `call_process_api`, the finally in `Queue.process_events` (a run that raised, was cancelled or lost its client, `/cancel` included), and the heartbeat's session cleanup. Two were missing and are added here. A generator called straight through `/gradio_api/run/{api_name}` yields once and is then dropped, because only the queue continues a generator and the event id that marks such a run is one the queue minted for it; an `event_id` arriving in the body of a direct call names no job, so the route drops it rather than let the request decide whether anybody will come back. That run used to leave its streams open, which cost segment bytes before this change and would cost a live ffmpeg process after it, one per abandoned run (twenty such calls measured at twenty processes and about 800 MiB before, none after). Those streams are flushed before they are ended, so the one chunk such a call returns comes out whole; ending them without the flush kept only what the encoder had emitted by then, 0.128 s of a 250 ms chunk and nothing at all of a 64 ms one, where `main` returned the chunk in full. And the first call of a run is out of reach of both the `call_process_api` handler and the queue's finally, because they look the run up through `app.iterators[event_id]`, which is assigned only after `process_api` returns. `process_api` now ends the run's streams itself when its streaming block raises; on later calls it and the callers' handlers both fire, and ending a stream twice is a no-op. Completing a run's streams ends every one of them in a `finally`, because a flush that raises would otherwise strand the streams after it and, with no event id, the caller's handler has no iterator to resolve the run from.
- The teardown is registered as the handle from `weakref.finalize`, which runs once and disarms itself, so `end_stream()` releases the encoder and nothing stays armed to fire a second time. What the unarmed handle covers is interpreter exit, not a dropped event stream: the session cleanup that discards a `MediaStream` ends it first, so the finalizer never gets there.
- There is still no idle timeout, so a slow generator is never killed mid-stream.
- The registry entry is not the encoder but a slot for it, made on the event loop before the encoding thread is dispatched. A cancel is what needs that. `utils.cancel_tasks` calls `task.cancel()`, which anyio's `to_thread.run_sync` does not defer: the `CancelledError` surfaces at once and the thread runs on, and can register its encoder after the coroutine that asked for it is gone. Measured with a 10 ms first chunk, a cancel 3 ms in caught its `CancelledError` 0.1 ms before the thread registered, and every cancel from there to the end of the chunk left one live ffmpeg that no teardown could reach, since no `MediaStream` existed to hang a finalize on. The thread now attaches its encoder to the slot under a lock, or closes the encoder if the slot was ended first, and `stream_output` ends the slot when anything raises on a first chunk before the encoder is attached, cancellation and exception alike, whether at the encode, at the fetch of a URL value or at the read of a file value. Whichever side finishes last does the release, and the same rig shows nothing left at any cancel timing. A first chunk that finds a slot already parked under its key ends it rather than overwriting it; keys are unique per run, so nothing reachable parks one, and this only keeps a bug elsewhere from orphaning a process silently.
- `merge_generated_values_into_output` (example caching) is a second, easily missed driver of `stream_output`, and it passed `output_id=""`. That was harmless while the id was only a URL, but as an encoder key it would have made every cached streaming example share one encoder and leak the process for the app's lifetime. It now uses a unique id per example and releases the encoder in a `finally`. It also flushes the tail, because that driver never finishes the stream and the frames still inside the encoder would be dropped. For the three yields of `test_caching_with_generators_and_streamed_output`, the cached example comes out at 3.181x its input where `main` gives 3.435x, the difference being the three priming frames `main` adds.
- Segment durations come from the emitted frame count rather than from the source chunk, and `#EXTINF` carries six decimals. Packed audio has no timestamps of its own, so a player derives each segment's start by accumulating those values, and three decimals cannot place a 1024-sample frame (0.0232199 s at 44.1 kHz) closely enough over a long stream.
- The encoder asks for its own resample rather than leaving one to ffmpeg. An AAC stream can only declare the rates in the ADTS table, and a frame is 1024 samples at whatever rate comes out, so a rate outside that table made every `#EXTINF` wrong: 20 kHz in claimed 1.1776 s per 23 frames where the media carried 1.0681 s, 10.3% out. Checked against the sampling-frequency index in the emitted frames for 8000, 20000, 40000, 44100 and 6000 Hz.
- A dead encoder now raises instead of going quiet. Nothing checked ffmpeg's exit code, so a process that died partway left a stream that simply stopped growing and a playlist that still got its `#EXT-X-ENDLIST`; the old per-chunk path raised out of pydub, so this would have been a step back. A deliberate `close()` is told apart from that. A chunk in flight when a disconnect or a cancel tore its stream down used to surface as `The audio encoder exited with code -9`, the SIGKILL `close()` itself had sent, and a `flush()` that killed an ffmpeg still running 5 s after its input ended raised about that code and dropped the frames it had already collected. After `close()`, `feed()` is a no-op and `take()` and `flush()` return what they have; the flush that has to kill returns its frames and logs a warning. `close()` also joins the reader thread before closing the pipe it reads from, since the reader loops on a cached fd and a closed fd number can be reused by another thread's file or socket.
- A wav chunk that declares a `data` size of 0, the usual placeholder when the length is not yet known, falls through to ffmpeg rather than being read as no audio at all. The stdlib `wave` module trusts the declared size, so such a chunk decoded to nothing and was dropped in silence; pydub sent it through ffmpeg, which reads to EOF. A chunk that really carries no samples, which a generator produces whenever a tick yields `(rate, np.zeros(0))`, then comes back from ffmpeg as nothing, and the decoder took nothing for a failure: `main` emitted a zero-length segment and played on, while this branch killed the run on any empty chunk after the first. The decoder now goes by ffmpeg's exit code, so an empty chunk produces no segment, which plays the same as `main`'s empty one, and a chunk that does fail carries ffmpeg's own message. The encoder runs at `-v error` as well, so a process that dies says why on the server's stderr.

- `end_stream()` runs `close()` synchronously on the event loop: a SIGKILL, a reap and a thread join. Measured at 1.1 ms idle and 3.5 ms with the reader mid-read; the multi-second ceiling the timeouts allow needs a process that ignores SIGKILL.
- The slot could live on `MediaStream` instead of in a module-level dict keyed by the playlist path, which would do away with the string key and the registry. That means building the `MediaStream` before the first `stream_output` and changing that method's signature, which Video and custom streaming components implement, so I have left it as a follow-up. Happy to do it here if you would rather.
- The flush at the end of a stream takes the encoder out of the registry and the slot before it hands it to a thread, and anyio's `to_thread.run_sync` yields to the loop once before the thread exists. A cancel landing in that gap used to leave the encoder held by nothing, one live ffmpeg per cancel for the life of the process; the await now closes the encoder on any exception. A cancel that lands after dispatch races the thread's own close, which is safe to repeat.
- `StreamingOutput` gains `flush_stream_output` and `end_stream_output`. Both are concrete no-ops on the base class, so Video and custom streaming components are unaffected, but it is a little more public surface than a bug fix usually adds and I am happy to reshape it. `Audio.covert_to_adts` goes away with the per-chunk path; nothing in the tree called it and it is not in the docs, but the name had no underscore.

What this does not cover, since the title could be read as broader than it is: the continuity guarantee is for wav and PCM input, which is what a numpy yield and `format="wav"` produce. A generator yielding mp3 or opus still has each chunk decoded on its own, so any decoder-side delay and non-frame-aligned split survives into the PCM the encoder then joins up. That is unchanged from `main` rather than a new gap, and the added test covers the wav path only.

Two smaller things this does not change, in case they come up in review: `MediaStream.playlist` and `MediaStream.segment_index` were dead fields and are already gone from `main`, and the float `#EXT-X-TARGETDURATION` was already fixed there too.

One part of the reported choppiness is outside this change. On a generator that produces audio in real time, hls.js also stalls, because the player is configured `maxBufferLength: 1, lowLatencyMode: true` (`js/audio/player/AudioPlayer.svelte`) and has no slack when playback outruns delivery. That is pre-existing and independent of the encoder; holding `play()` until roughly a second is buffered removes it in my measurements, at the cost of the startup latency that config exists to minimise, so it is a product call rather than something to slip in here.

That the stall is the player's rather than the server's is worth showing, because it was not true of `main` and would not have been true of a first draft of this change either. Wall time per `stream_output` call at 16 kHz mono, and the supply rate it implies for a generator that produces instantly:

| chunk | `main` | this PR |
|---|---|---|
| 160 (10 ms) | 162.9 ms, 0.06x | 9.3 ms, 1.08x |
| 320 (20 ms) | 160.4 ms, 0.12x | 8.3 ms, 2.42x |
| 512 (32 ms) | 160.9 ms, 0.20x | 7.2 ms, 4.42x |
| 1024 (64 ms) | 160.6 ms, 0.40x | 4.8 ms, 13.39x |
| 4000 (250 ms) | 159.1 ms, 1.57x | 2.6 ms, 94.47x |

`main` costs a flat 160 ms per chunk whatever its size, being two pydub ffmpeg spawns, so below 250 ms chunks it cannot keep up with a real-time generator at all and no player config could hide that. This encoder is above real time at every size.

Closes: #11733

## Before / after Spaces

- [Before fix](https://huggingface.co/spaces/hysts-debug/gradio-11733-before)
- [After fix](https://huggingface.co/spaces/hysts-debug/gradio-11733-after)

Both run the same `app.py`: upload a file or pick a built-in signal (a frequency sweep, a steady tone, a click train), stream it back through `gr.Audio(streaming=True)` in chunks of a chosen length, then compare the mel spectrograms of the source and of the stream as served, on one time scale. The before Space is gradio 6.26.0 from PyPI; the after Space installs the preview wheel for this PR's head commit, pinned by SHA. The release also predates #13808, so on the before Space reload the page between runs (the app says so).

## Testing

Backend tests cover the continuity property directly: eight chunks through `stream_output` produce one ADTS stream with a single priming frame, and the playlist's `#EXTINF` sum matches what the segments decode to, at 16 kHz and at 20 kHz, where the frames come out at 22.05 kHz and the durations have to follow them. The lifecycle has its own tests: a first chunk cancelled before and after its encoder exists, a first call that raises after one output has opened a stream, a stream that opens on a `None` and gets its audio later, a chunk in flight when its encoder is closed, a flush that has to kill the encoder (and that it keeps the frames it had), a flush cancelled before its thread runs, a first chunk whose file is missing, an encoder whose reader thread cannot start, an empty chunk mid-stream, a generator called through the run route, whose one chunk has to come back whole whether or not the body carries an `event_id`, and a run with two streaming outputs where the first one's flush raises. Each was confirmed to fail without the change it pins. The `requires_ffmpeg` marker moved from `test_blocks.py` to `conftest.py` so the audio component tests can use it.

What I could and could not verify by hand, since the design leans on browser behaviour:

- **Chromium**: captured playback of a real app through hls.js. The per-chunk gaps are gone. That capture predates the fixes in the later commits here, but the segment bytes for its configuration are unchanged since, which the gap measurement above still reproduces.
- **Firefox**: no audio device on my machine, so its `AudioContext` never leaves `suspended`. Covered indirectly instead, by checking decoder state across segments with WebCodecs and by checking append contiguity through MSE, both cross-validated against Chromium.
- **Safari and native HLS**: untested. Worth noting that `main` does not emit `#EXT-X-DISCONTINUITY` for `.aac`, and this PR does not change that, so native-HLS players see the same playlist structure as before.
- **Windows**: untested. The encoder reads the ffmpeg pipe with `os.read` on a raw fd.

## AI Disclosure

We encourage the use of AI tooling in creating PRs, but the any non-trivial use of AI needs be disclosed. E.g. if you used Claude to write a first draft, you should mention that. Trivial tab-completion doesn't need to be disclosed. You should self-review all PRs, especially if they were generated with AI.

- [x] I used AI to investigate the root cause and implement the fix.
- [ ] I did not use AI

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small.

Not adhering to this guideline will result in the PR being closed.

## Testing and Formatting Your Code

1. PRs will only be merged if tests pass on CI. We recommend at least running the backend tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the backed tests: `bash scripts/run_backend_tests.sh`

2. Please run these bash scripts to automatically format your code: `bash scripts/format_backend.sh`, and (if you made any changes to non-Python files) `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/audio_stream_encoder.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            "-ar", str(sample_rate), "-ac", str(channels),
+            "pipe:1",
+        ],
+        input=data,
+        capture_output=True,
+        check=False,
+    )  # fmt: skip
+    if not result.stdout:
+        raise RuntimeError("Could not decode the streamed audio chunk.")
+    return result.stdout
```
[the remark this answers is not in the document]
Comment (hysts) [reply gradio/audio_stream_encoder.py:108]: This path is not reachable without the clearer error firing first. `_ffmpeg_decode` is only called from the branch of `decode_to_pcm` where both `sample_rate` and `channels` are given, and that only happens for the second and later chunks of a stream, after `AacStreamEncoder.__init__` has already run `shutil.which("ffmpeg")` and raised the explicit `RuntimeError` if it is missing. The first chunk goes through `_read_wav_pcm` or pydub, neither of which reaches this function.

So a missing ffmpeg surfaces as "Streaming audio output requires `ffmpeg` to be installed and on PATH." before any chunk is decoded here. Leaving it as is rather than adding a second check for a state the stream cannot be in.

File: gradio/components/audio.py [hunk 2]
```diff
@@ -358,22 +375,31 @@ async def stream_output(
...
+        )
+        return chunk, output_file
+
+    async def flush_stream_output(self, output_id: str) -> MediaStreamChunk | None:
+        encoder = _stream_encoders.pop(output_id, None)
+        if encoder is None:
+            return None
+        return await anyio.to_thread.run_sync(
+            lambda: _segment_from_frames(encoder, encoder.flush())
+        )
```
[the remark this answers is not in the document]
Comment (hysts) [reply gradio/components/audio.py]: Fixed in 28dfdbaa7. The process itself was already waited on inside `flush()`, but you are right that the stdout pipe was left to the collector. Rather than adding the `close()` call at the `flush_stream_output` call site, `flush()` now hands off to `close()` at the end, since it is the terminal call and every caller that flushes then drops the encoder gets the descriptors released.

File: gradio/audio_stream_encoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hysts) [reply gradio/audio_stream_encoder.py:108]: Follow-up on the same subject, since round two raised a nearby path (`decode_to_pcm` at the pydub fallback) rather than this one.

That fallback only runs for the *first* chunk of a stream, and only when the bytes are not 16-bit PCM wav, so it is reached before `AacStreamEncoder` exists and a missing ffmpeg would indeed surface as pydub's own error. It is unchanged behaviour, though: `main` calls `AudioSegment.from_file` plus `export(format="adts")` on every chunk with no ffmpeg check at all, so this PR narrows that path rather than widening it, and adds an explicit message for the common case. Adding a second `shutil.which` guard in front of pydub would be a change to error reporting on a path this PR is not about, so I would rather leave it out here.

The other point from that round was worth taking: cleanup callbacks that raise in `end_stream()` are now logged at debug level with `exc_info`, matching what this module already does for the history writer, in c01883795.

File: test/conftest.py [hunk 3]
```diff
@@ -7,13 +7,26 @@
...
     config.addinivalue_line(
         "markers", "flaky: mark test as flaky. Failure will not cause te"
     )
     config.addinivalue_line("markers", "serial: mark test as serial")
+    config.addinivalue_line(
+        "markers", "requires_ffmpeg: skip the test when ffmpeg is not on PATH"
+    )
+
+
+def pytest_collection_modifyitems(items):
```
[the remark this answers is not in the document]
Comment (hysts) [reply test/conftest.py:23]: A pytest hook implementation may take any subset of the hookspec's arguments; pluggy passes them by name, so `(items)` alone is a valid signature and is called without error. It has also run as is: the CI job for 243fbc34f collected and executed the tests carrying this marker, which a `TypeError` during collection would have prevented. Leaving this open in case a maintainer prefers the full signature regardless.

File: test/components/test_audio.py [hunk 4]
```diff
@@ -9,11 +15,220 @@
...
+        assert stream_id not in _stream_encoders
+
+        proceed.set()
+        deadline = time.monotonic() + 5
+        while not (encoders and encoders[0].process.poll() is not None):
+            assert time.monotonic() < deadline, "the encoder was never released"
+            await asyncio.sleep(0.005)
+
+    @pytest.mark.requires_ffmpeg
+    @pytest.mark.requires_ffmpeg
```
[the remark this answers is not in the document]
Comment (hysts) [reply test/components/test_audio.py]: Yes, and the cause was one step further back: the test above this one was inserted between this marker and `test_a_closed_encoder_stays_quiet`, which lost it. Both put right in 20da5bcfd.

**Comment by abidlabs** [comment]
Two lifecycle edge cases I could reproduce:

1. A direct `/run` caller can supply an [`event_id`](https://github.com/gradio-app/gradio/blob/a270be0fa9c6e61bf137e229fcbdd21aa19eb586/gradio/data_classes.py#L91-L94). [`call_process_api()` then treats it as queue-owned](https://github.com/gradio-app/gradio/blob/a270be0fa9c6e61bf137e229fcbdd21aa19eb586/gradio/route_utils.py#L431-L444), skips `_finish_run_streams()`, and leaves the iterator plus live ffmpeg referenced indefinitely because no queue job owns that ID. Directness should come from trusted route context rather than the request body.

2. [`_finish_run_streams()` stops at the first flush exception](https://github.com/gradio-app/gradio/blob/a270be0fa9c6e61bf137e229fcbdd21aa19eb586/gradio/blocks.py#L2167-L2187). With multiple streamed Audio outputs, if the first flush raises, later streams are never ended; for an event-less direct call, the exception handler also cannot recover the local iterator/run. Please ensure every remaining stream is ended in a `finally` (or preserve the run for cleanup).

**Review by abidlabs — APPROVED** [review]
Great PR and benchmarking @hysts! Overall, this approach looks solid and I can confirm the improvements in performance. I think optimizing streaming is going to be a big part of Gradio 7 so really great that we already have this in place.

I put a couple of minor comments above. Also I was thinking we should do the same as part of `gr.Video` streaming as well, right? Could be done as a follow up PR

**Comment by hysts** [comment]
Thanks for the review @abidlabs, and for catching both of these. Fixed in 83f6f0731.

Agreed on `gr.Video`. It has the same per-chunk encode, and it is too big to fold in here, so I will do it as a follow-up PR.

Will merge once CI is green.

**Review by brownkadarius81-hue — APPROVED** [review]
